cluster/manager.go

```diff
@@ -0,0 +1,330 @@
+	c.RLock()
+	defer c.RUnlock()
+	return c.node
+}
+
+func (c *ClusterManager) PeersList() []Node {
+	c.RLock()
+	list := make([]Node, len(c.Peers), len(c.Peers))
+	i := 0
+	for _, p := range c.Peers {
+		list[i] = p
+		i++
+	}
+	c.RUnlock()
+	return list
+}
+
+// report the cluster stats every time there is a change to the cluster state.
+// it is assumed that the lock is acquired before calling this method.
+func (c *ClusterManager) clusterStats() {
+	primReady := 0
+	primNotReady := 0
+	secReady := 0
+	secNotReady := 0
+	partitions := make(map[int32]int)
+	for _, p := range c.Peers {
+		if p.Primary {
+			if p.IsReady() {
+				primReady++
+			} else {
+				primNotReady++
+			}
+		} else {
+			if p.IsReady() {
+				secReady++
+			} else {
+				secNotReady++
+			}
+		}
+		for _, partition := range p.Partitions {
+			partitions[partition]++
+		}
+	}
+
+	totalPrimaryReady.Set(primReady)
+	totalPrimaryNotReady.Set(primNotReady)
+	totalSecondaryReady.Set(secReady)
+	totalSecondaryNotReady.Set(secNotReady)
+
+	totalPartitions.Set(len(partitions))
+}
+
+func (c *ClusterManager) NotifyJoin(node *memberlist.Node) {
+	eventsJoin.Inc()
+	c.Lock()
+	defer c.Unlock()
+	if len(node.Meta) == 0 {
+		return
+	}
+	log.Info("CLU manager: Node %s with address %s has joined the cluster", node.Name, node.Addr.String())
+	peer := Node{}
+	err := json.Unmarshal(node.Meta, &peer)
+	if err != nil {
+		log.Error(3, "CLU manager: Failed to decode node meta from %s: %s", node.Name, err.Error())
+		unmarshalErrJoin.Inc()
+		return
+	}
+	peer.RemoteAddr = node.Addr.String()
+	if peer.Name == c.node.Name {
+		peer.local = true
+	}
+	c.Peers[node.Name] = peer
+	c.clusterStats()
+}
+
+func (c *ClusterManager) NotifyLeave(node *memberlist.Node) {
+	eventsLeave.Inc()
+	c.Lock()
+	defer c.Unlock()
+	log.Info("CLU manager: Node %s has left the cluster", node.Name)
+	delete(c.Peers, node.Name)
+	c.clusterStats()
+}
+
+func (c *ClusterManager) NotifyUpdate(node *memberlist.Node) {
+	eventsUpdate.Inc()
+	c.Lock()
+	defer c.Unlock()
+	if len(node.Meta) == 0 {
+		return
+	}
+	peer := Node{}
+	err := json.Unmarshal(node.Meta, &peer)
+	if err != nil {
+		log.Error(3, "CLU manager: Failed to decode node meta from %s: %s", node.Name, err.Error())
+		unmarshalErrUpdate.Inc()
+		// if the node is known, lets mark it as notReady until it starts sending valid data again.
+		if p, ok := c.Peers[node.Name]; ok {
+			p.State = NodeNotReady
+			p.StateChange = time.Now()
+			// we dont set Updated as we dont want the NotReady state to propagate incase we are the only node
+			// that got bad data.
+			c.Peers[node.Name] = p
+		}
+		return
+	}
+	peer.RemoteAddr = node.Addr.String()
+	if peer.Name == c.node.Name {
+		peer.local = true
+	}
+	c.Peers[node.Name] = peer
+	log.Info("CLU manager: Node %s at %s has been updated - %s", node.Name, node.Addr.String(), node.Meta)
+	c.clusterStats()
+}
+
+func (c *ClusterManager) BroadcastUpdate() {
+	if c.list != nil {
+		go c.list.UpdateNode(time.Second)
```

Comment:
This will only fail if connectivity to all nodes is lost. It is unlikely that will happen without this node being dropped from the cluster.  When it re-joins it will do a full re-sync.  

Additionally, full state re-syncs are performed periodically (every 30seconds by default). Any changes not propagated via a NotifyUpdate message will be picked up via the complete state syncs, handled by the MergeRemoteState method.